# Phase 1 - Week 3 - Day 1 AM - Recommender System

# A. Content Based Filtering

Content Based Filtering usses item metadata as the basis for the recommendation

We'll try to make a simple content-based recommender system using cosine similarity. We measure the similarity between two items. Conceptually, we will recommend a user some items based on the simility of items that they like previously.

To remind you, cosine similarity is a 'distance' measurement of two vectors

<img src="https://softscients.com/wp-content/uploads/2020/03/2.-Cara-Menghitung-Cosine-similarity.png"></img>


<img src="https://www.researchgate.net/profile/Said-Salloum/publication/345471138/figure/fig2/AS:955431962808321@1604804139868/Cosine-similarity-formula.png"></img>

To start, we use Pandas for data loading and preprocessing and Numpy for linear algebra calculation. In this case, we will make a movie recommendation system.

In [1]:
# Import library

import pandas as pd
import numpy as np

In [2]:
# Data loading

movie = pd.read_csv('https://raw.githubusercontent.com/FTDS-learning-materials/phase-1/refs/heads/v2.4/w3/P1W3D1AM%20-%20Recommender%20System%20-%20movies.csv').dropna()
movie.head()

,index,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,...,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,cast,crew,director
0,0,237000000,Action Adventure Fantasy Science Fiction,http://www.avatarmovie.com/,19995,culture clash future space war space colony so...,en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,...,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,Sam Worthington Zoe Saldana Sigourney Weaver S...,"[{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...",James Cameron
1,1,300000000,Adventure Fantasy Action,http://disney.go.com/disneypictures/pirates/,285,ocean drug abuse exotic island east india trad...,en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,...,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,Johnny Depp Orlando Bloom Keira Knightley Stel...,"[{'name': 'Dariusz Wolski', 'gender': 2, 'depa...",Gore Verbinski
2,2,245000000,Action Adventure Crime,http://www.sonypictures.com/movies/spectre/,206647,spy based on novel secret agent sequel mi6,en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,...,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466,Daniel Craig Christoph Waltz L\u00e9a Seydoux ...,"[{'name': 'Thomas Newman', 'gender': 2, 'depar...",Sam Mendes
3,3,250000000,Action Crime Drama Thriller,http://www.thedarkknightrises.com/,49026,dc comics crime fighter terrorist secret ident...,en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,...,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106,Christian Bale Michael Caine Gary Oldman Anne ...,"[{'name': 'Hans Zimmer', 'gender': 2, 'departm...",Christopher Nolan
4,4,260000000,Action Adventure Science Fiction,http://movies.disney.com/john-carter,49529,based on novel mars medallion space travel pri...,en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,...,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124,Taylor Kitsch Lynn Collins Samantha Morton Wil...,"[{'name': 'Andrew Stanton', 'gender': 2, 'depa...",Andrew Stanton


In [3]:
# Get info

movie.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1432 entries, 0 to 4796
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   index                 1432 non-null   int64  
 1   budget                1432 non-null   int64  
 2   genres                1432 non-null   object 
 3   homepage              1432 non-null   object 
 4   id                    1432 non-null   int64  
 5   keywords              1432 non-null   object 
 6   original_language     1432 non-null   object 
 7   original_title        1432 non-null   object 
 8   overview              1432 non-null   object 
 9   popularity            1432 non-null   float64
 10  production_companies  1432 non-null   object 
 11  production_countries  1432 non-null   object 
 12  release_date          1432 non-null   object 
 13  revenue               1432 non-null   int64  
 14  runtime               1432 non-null   float64
 15  spoken_languages      1432

## A.1. Based on `genres` Only

To simplify our system, we will use genres only for the vector elements. Remember that, cosine similarity needs vectors to do the calculation so we have to extract the vector from the genre data.

To extract the vector, we use one-hot encoding technique (a technique that labeling of the existance of a category), which the illustration represented by image below:

<img src="https://i.imgur.com/mtimFxh.png"></img>

since each movie has more than one genres, we have to do an extra preprocessing.

In [4]:
# Get all genres

genres = ' '
for g in movie['genres']:
  genres += g+' '

genres = list(set(genres.split(' ')))[1:]

In [5]:
# Show all genres

genres

['Drama',
 'Documentary',
 'Adventure',
 'Western',
 'Action',
 'Crime',
 'Horror',
 'Science',
 'Movie',
 'Fiction',
 'Family',
 'Music',
 'Romance',
 'Foreign',
 'Mystery',
 'History',
 'Fantasy',
 'Comedy',
 'Animation',
 'TV',
 'Thriller',
 'War']

In [6]:
# Create a matrix to map films by genre

gen_mv = [[] for i in range(len(genres))]

for dat in movie['genres']:
  for i,g in enumerate(genres):
    if g in dat.split(' '):
      gen_mv[i].append(1)
    else:
      gen_mv[i].append(0)

In [7]:
# Displays the matrix that maps films by genre.

gen_mv_dat = pd.DataFrame(np.array(gen_mv).T, columns=genres)
gen_mv_dat

,Drama,Documentary,Adventure,Western,Action,Crime,Horror,Science,Movie,Fiction,...,Romance,Foreign,Mystery,History,Fantasy,Comedy,Animation,TV,Thriller,War
0,0,0,1,0,1,0,0,1,0,1,...,0,0,0,0,1,0,0,0,0,0
1,0,0,1,0,1,0,0,0,0,0,...,0,0,0,0,1,0,0,0,0,0
2,0,0,1,0,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,1,0,0,0,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0
4,0,0,1,0,1,0,0,1,0,1,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1427,1,0,0,0,1,0,0,0,0,0,...,0,0,0,0,0,1,0,0,0,0
1428,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,1,0,0,0,0
1429,0,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,1,0,0,0,0
1430,0,0,0,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [8]:
# Combine the matrix with the movie name

title_df = movie[['original_title']].reset_index(drop=True)
movie_vector = pd.concat([title_df,gen_mv_dat],axis=1)
movie_vector.set_index('original_title',inplace=True)
movie_vector

,Drama,Documentary,Adventure,Western,Action,Crime,Horror,Science,Movie,Fiction,...,Romance,Foreign,Mystery,History,Fantasy,Comedy,Animation,TV,Thriller,War
original_title,,,,,,,,,,,,,,,,,,,,,
Avatar,0,0,1,0,1,0,0,1,0,1,...,0,0,0,0,1,0,0,0,0,0
Pirates of the Caribbean: At World's End,0,0,1,0,1,0,0,0,0,0,...,0,0,0,0,1,0,0,0,0,0
Spectre,0,0,1,0,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
The Dark Knight Rises,1,0,0,0,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0
John Carter,0,0,1,0,1,0,0,1,0,1,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Down Terrace,1,0,0,0,1,0,0,0,0,0,...,0,0,0,0,0,1,0,0,0,0
Clerks,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,1,0,0,0,0
Dry Spell,0,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,1,0,0,0,0


Voila! we have the movie vectors which represent each movie's genres. Next, we define the cosine similarity function to ease our similarity calculation.

In [9]:
# Define cosine similarity function to get the similarity between 2 movies

def cosine_sim(vect1,vect2):
  norm_1 = np.linalg.norm(vect1)
  norm_2 = np.linalg.norm(vect2)

  cos_sim = (vect1 @ vect2) / (norm_1 * norm_2)
  return cos_sim

We want to test the function for Avatar and Tin Can Man. In fact, Avatar and Tin Can Man have different genres

In [10]:
# Show the movie `Avatar` and its genre

movie[movie['original_title']=='Avatar'][['original_title','genres']]

,original_title,genres
0,Avatar,Action Adventure Fantasy Science Fiction


In [11]:
# Show the movie `Tin Can Man` and its genre

movie[movie['original_title']=='Tin Can Man'][['original_title','genres']]

,original_title,genres
4791,Tin Can Man,Horror


In [12]:
# Get the similarity

cosine_sim(movie_vector.loc['Avatar'], movie_vector.loc['Tin Can Man'])

np.float64(0.0)

The cosine similarity of both movies is zero, which is there is no similarity between them.

Imagine that you really love `Man of Steel` and our system will recommend you 5 movies that similar to `Man of Steel`.

In [13]:
# Function to find top n films that are similar to the input film

def recsys(movie, top_N):
  cossim = pd.Series([cosine_sim(movie_vector.loc[movie], x) for x in movie_vector.values],index=movie_vector.index).drop(index=movie)
  print(f'You like {movie}, so based on our recommender system, We recommend you to watch:')
  for i, mv in enumerate(cossim.sort_values(ascending=False)[:top_N].index):
    print(f'{i+1}. {mv} - similarity {cossim[mv]}')

  return cossim

In [14]:
# Show top n movies similar to `Man of Steel`

cossim_result = recsys('Man of Steel',5)

You like Man of Steel, so based on our recommender system, We recommend you to watch:
1. X-Men: Days of Future Past - similarity 0.9999999999999998
2. Jupiter Ascending - similarity 0.9999999999999998
3. Avatar - similarity 0.9999999999999998
4. The Wolverine - similarity 0.9999999999999998
5. Suicide Squad - similarity 0.9128709291752769


In [15]:
# Show cosine similarity result

cossim_result.sort_values(ascending=False)

,0
original_title,
X-Men: Days of Future Past,1.000000
Jupiter Ascending,1.000000
Avatar,1.000000
The Wolverine,1.000000
Suicide Squad,0.912871
...,...
One Night with the King,0.000000
What's Your Number?,0.000000
Sin City: A Dame to Kill For,0.000000


## A.2 - Based on `overview`

In [16]:
# Example of a film description

movie['original_title'][0], movie['overview'][0]

('Avatar',
 'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.')

To preprocess the text into numbers, we can approach by use a tf-idf vectorizer.

While this approach is more commonly used on a text corpus, it possesses some interesting properties that will be useful in order to obtain a vector representation of the data. The expression is defined as follows:

![img](https://raw.githubusercontent.com/AlexanderNixon/Machine-learning-reads/b47791834906c152411dfaf5f5c2035aebd2157d//images/Content_based_recommenders/tfidf.jpg)

Where we have the product of the term frequency, i.e. the amount of times a given term (words) occurs in a document (overview of a movie), times the right side factor, which basically scales the term frequency depending on the amount of times a given term appears in all documents (movies).

In [17]:
# Word embedding using TF-IDF

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(movie['overview'])
tfidf_matrix

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 37981 stored elements and shape (1432, 10886)>

**Notes **

- `1432` : Number of movies
- `10886` : Total number of unique words in all documents combined

In [18]:
# Calculate cosine similarity between movies

cosine_sim = cosine_similarity(tfidf_matrix)

df = pd.DataFrame(cosine_sim, index=movie['original_title'], columns=movie['original_title'])
df.sample(5, axis=1).round(2)

original_title,The In Crowd,Савва. Сердце воина,Captain America: The First Avenger,The Lincoln Lawyer,Old School
original_title,,,,,
Avatar,0.00,0.00,0.00,0.00,0.00
Pirates of the Caribbean: At World's End,0.00,0.01,0.03,0.00,0.00
Spectre,0.00,0.00,0.03,0.00,0.00
The Dark Knight Rises,0.04,0.00,0.03,0.00,0.00
John Carter,0.00,0.02,0.07,0.00,0.00
...,...,...,...,...,...
Down Terrace,0.00,0.00,0.01,0.03,0.00
Clerks,0.00,0.00,0.00,0.00,0.00
Dry Spell,0.00,0.08,0.00,0.00,0.00


In [19]:
# Show all cosine similarity matrix

df

original_title,Avatar,Pirates of the Caribbean: At World's End,Spectre,The Dark Knight Rises,John Carter,Spider-Man 3,Tangled,Avengers: Age of Ultron,Harry Potter and the Half-Blood Prince,Batman v Superman: Dawn of Justice,...,Tiger Orange,Tiny Furniture,The Signal,Raymond Did It,The Last Waltz,Down Terrace,Clerks,Dry Spell,Tin Can Man,Primer
original_title,,,,,,,,,,,,,,,,,,,,,
Avatar,1.000000,0.000000,0.0,0.026412,0.000000,0.029903,0.000000,0.039419,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.042604
Pirates of the Caribbean: At World's End,0.000000,1.000000,0.0,0.000000,0.033908,0.000000,0.000000,0.024137,0.000000,0.000000,...,0.000000,0.009465,0.000000,0.000000,0.0,0.000000,0.0,0.019407,0.016146,0.000000
Spectre,0.000000,0.000000,1.0,0.000000,0.000000,0.000000,0.000000,0.031617,0.028132,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.0,0.000000,0.000000,0.000000
The Dark Knight Rises,0.026412,0.000000,0.0,1.000000,0.010363,0.006066,0.013311,0.024691,0.018978,0.132153,...,0.019367,0.005529,0.013833,0.024989,0.0,0.021553,0.0,0.048796,0.000000,0.000000
John Carter,0.000000,0.033908,0.0,0.010363,1.000000,0.000000,0.009359,0.040568,0.000000,0.018681,...,0.000000,0.000000,0.000000,0.000000,0.0,0.012571,0.0,0.000000,0.000000,0.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Down Terrace,0.000000,0.000000,0.0,0.021553,0.012571,0.000000,0.008907,0.000000,0.000000,0.015714,...,0.000000,0.000000,0.000000,0.000000,0.0,1.000000,0.0,0.000000,0.000000,0.000000
Clerks,0.000000,0.000000,0.0,0.000000,0.000000,0.011578,0.018727,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,1.0,0.000000,0.000000,0.000000
Dry Spell,0.000000,0.019407,0.0,0.048796,0.000000,0.000000,0.000000,0.056511,0.000000,0.000000,...,0.000000,0.042219,0.000000,0.000000,0.0,0.000000,0.0,1.000000,0.019196,0.000000


In [20]:
# Show top 5 movies similar to `Avatar`

df['Avatar'].drop(index='Avatar').sort_values(ascending=False).iloc[:5]

,Avatar
original_title,
Apollo 18,0.168557
The Matrix,0.132872
Hanna,0.105987
Semi-Pro,0.087084
Aliens vs Predator: Requiem,0.078278


In [21]:
# Function to find top 5 films that are similar to the input film

def sorting(mv):
  tmp = df[mv].drop(index=mv).sort_values(ascending=False).iloc[:5]
  print(f'You like {mv}, so based on our recommender system, We recommend you to watch:')
  for i,mv in enumerate(tmp.index):
    print(f'{i+1}. {mv}')

In [22]:
# Show top 5 movies similar to `Man of Steel`

sorting('Man of Steel')

You like Man of Steel, so based on our recommender system, We recommend you to watch:
1. Prometheus
2. Ong Bak 2
3. Midnight Special
4. War Horse
5. Astro Boy


In [23]:
# Syntax example for searching movies that include a certain word

df.index[df.index.str.contains('jurassic', case=False)]

Index(['Jurassic World', 'Jurassic Park'], dtype='object', name='original_title')

In [24]:
# Show top 5 movies similar to `Jurassic Park`

sorting('Jurassic Park')

You like Jurassic Park, so based on our recommender system, We recommend you to watch:
1. Jurassic World
2. Walking With Dinosaurs
3. Adventureland
4. The Good Dinosaur
5. Yogi Bear


# B. Collaborative Filtering

Collaborative Filtering filter out items that a user might like on the basis of reactions by similar users.

We will develop a simple recommender system using memory-based collaboratice filtering.

Memory-based collaborative filtering (also known as neighborhood-based collaborative filtering) is a recommendation technique that makes predictions for users or items based on the preferences of similar users or items. It can be divided into two main categories: user-based collaborative filtering and item-based collaborative filtering. However, we will focus on user-based (user-user) technique.

In [25]:
# Data loading

books = pd.read_csv('https://raw.githubusercontent.com/FTDS-learning-materials/phase-1/refs/heads/v2.4/w3/P1W3D1AM%20-%20Recommender%20System%20-%20books.csv')
ratings = pd.read_csv('https://raw.githubusercontent.com/FTDS-learning-materials/phase-1/refs/heads/v2.4/w3/P1W3D1AM%20-%20Recommender%20System%20-%20ratings.csv')

In [26]:
# Display data `books`

books.head()

,book_id,goodreads_book_id,best_book_id,work_id,books_count,isbn,isbn13,authors,original_publication_year,original_title,...,ratings_count,work_ratings_count,work_text_reviews_count,ratings_1,ratings_2,ratings_3,ratings_4,ratings_5,image_url,small_image_url
0,1,2767052,2767052,2792775,272,439023483,9.780439e+12,Suzanne Collins,2008.0,The Hunger Games,...,4780653,4942365,155254,66715,127936,560092,1481305,2706317,https://images.gr-assets.com/books/1447303603m...,https://images.gr-assets.com/books/1447303603s...
1,2,3,3,4640799,491,439554934,9.780440e+12,"J.K. Rowling, Mary GrandPré",1997.0,Harry Potter and the Philosopher's Stone,...,4602479,4800065,75867,75504,101676,455024,1156318,3011543,https://images.gr-assets.com/books/1474154022m...,https://images.gr-assets.com/books/1474154022s...
2,3,41865,41865,3212258,226,316015849,9.780316e+12,Stephenie Meyer,2005.0,Twilight,...,3866839,3916824,95009,456191,436802,793319,875073,1355439,https://images.gr-assets.com/books/1361039443m...,https://images.gr-assets.com/books/1361039443s...
3,4,2657,2657,3275794,487,61120081,9.780061e+12,Harper Lee,1960.0,To Kill a Mockingbird,...,3198671,3340896,72586,60427,117415,446835,1001952,1714267,https://images.gr-assets.com/books/1361975680m...,https://images.gr-assets.com/books/1361975680s...
4,5,4671,4671,245494,1356,743273567,9.780743e+12,F. Scott Fitzgerald,1925.0,The Great Gatsby,...,2683664,2773745,51992,86236,197621,606158,936012,947718,https://images.gr-assets.com/books/1490528560m...,https://images.gr-assets.com/books/1490528560s...


In [27]:
# Get the info of `books`

books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 23 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   book_id                    10000 non-null  int64  
 1   goodreads_book_id          10000 non-null  int64  
 2   best_book_id               10000 non-null  int64  
 3   work_id                    10000 non-null  int64  
 4   books_count                10000 non-null  int64  
 5   isbn                       9300 non-null   object 
 6   isbn13                     9415 non-null   float64
 7   authors                    10000 non-null  object 
 8   original_publication_year  9979 non-null   float64
 9   original_title             9415 non-null   object 
 10  title                      10000 non-null  object 
 11  language_code              8916 non-null   object 
 12  average_rating             10000 non-null  float64
 13  ratings_count              10000 non-null  int6

In [28]:
# Display data `ratings`

ratings.head()

,user_id,book_id,rating
0,1,258,5
1,2,4081,4
2,2,260,5
3,2,9296,5
4,2,2318,3


To ease our calculation, we need to transform the ratings dataframe into user-book ratings matrix

In [29]:
# Create a `user-book ratings` matrix

ratings_matrix = pd.pivot_table(ratings, values='rating', index='user_id', columns='book_id')
ratings_matrix.head()

book_id,1,2,3,4,5,7,8,9,10,11,...,9984,9985,9986,9988,9990,9991,9995,9997,9998,10000
user_id,,,,,,,,,,,,,,,,,,,,,
1,NaN,NaN,NaN,5.0,NaN,NaN,NaN,NaN,4.0,5.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,3.0,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,5.0,NaN,4.0,4.0,4.0,4.0,NaN,5.0,4.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [30]:
# Get the shape of `user-book ratings` matrix

ratings_matrix.shape

(3788, 5743)

**We have `3,788` users and `5,743` books.**

- Initially, the number of books we had in the dataset was 10,000.
- However, after performing a pivot table, it turned out that not all books had been rated.
- Only 5,743 books had received ratings from readers.
- There are 4,257 books that have not yet received any ratings.

In [31]:
# Get all the ratings from user_id = 1

ratings[ratings['user_id'] == 1].sort_values('book_id', ascending=True)

,user_id,book_id,rating
230502,1,4,5
230500,1,10,4
230489,1,11,5
230513,1,13,4
230512,1,32,4
230496,1,33,4
230495,1,35,5
230509,1,36,4
230507,1,42,3
230517,1,43,4


Normalize/Scaling the data to ease the similarity calculation

In [32]:
# Normalization

normalized_ratings_matrix = ratings_matrix.divide(ratings_matrix.mean(axis=1), axis=0).fillna(0)
normalized_ratings_matrix.head()

book_id,1,2,3,4,5,7,8,9,10,11,...,9984,9985,9986,9988,9990,9991,9995,9997,9998,10000
user_id,,,,,,,,,,,,,,,,,,,,,
1,0.0,0.000000,0.0,1.397516,0.000000,0.000000,0.000000,0.0,1.118012,1.397516,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,0.000000,0.0,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,0.0,0.000000,0.0,1.791667,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.0,1.326733,0.0,1.061386,1.061386,1.061386,1.061386,0.0,1.326733,1.061386,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
6,0.0,0.000000,0.0,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [33]:
# Average rating from each user

ratings_matrix.mean(axis=1).head(5)

,0
user_id,
1,3.577778
2,4.363636
3,1.674419
4,3.768657
6,4.000000


Calculate the similarity among users by their rating on certain book

In [34]:
# Define cosine similarity function to get the similarity between 2 movies

def cosine_sim(vect1,vect2):
  norm_1 = np.linalg.norm(vect1)
  norm_2 = np.linalg.norm(vect2)

  cos_sim = (vect1 @ vect2) / (norm_1 * norm_2)
  return cos_sim

# Calculate cosine similarity between user_id=1 and user_id=2
cosine_sim(normalized_ratings_matrix.loc[1], normalized_ratings_matrix.loc[2])

np.float64(0.04333895421846145)

In [35]:
# Calculate cosine similarity between users

cossim = cosine_similarity(normalized_ratings_matrix)
df = pd.DataFrame(cossim, index=normalized_ratings_matrix.index, columns=normalized_ratings_matrix.index)
df.sample(5, axis=1).round(2)

user_id,2020,2688,1829,5263,2684
user_id,,,,,
1,0.00,0.00,0.05,0.00,0.03
2,0.00,0.00,0.01,0.00,0.04
3,0.00,0.02,0.03,0.00,0.05
4,0.05,0.07,0.12,0.02,0.08
6,0.00,0.00,0.00,0.00,0.00
...,...,...,...,...,...
5760,0.00,0.01,0.01,0.01,0.07
5766,0.00,0.06,0.07,0.00,0.12
5895,0.00,0.00,0.00,0.10,0.13


In [36]:
# Show all cosine similarity matrix between users

df

user_id,1,2,3,4,6,7,8,9,10,11,...,5748,5749,5751,5757,5758,5760,5766,5895,6840,7582
user_id,,,,,,,,,,,,,,,,,,,,,
1,1.000000,0.043339,0.082742,0.214442,0.0,0.047795,0.081802,0.176935,0.214043,0.127841,...,0.025984,0.210911,0.049473,0.067586,0.056650,0.112011,0.238505,0.0,0.024889,0.000000
2,0.043339,1.000000,0.019642,0.066366,0.0,0.000000,0.044385,0.040322,0.049326,0.082677,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.044879,0.0,0.000000,0.000000
3,0.082742,0.019642,1.000000,0.108787,0.0,0.060331,0.079091,0.090668,0.059294,0.025055,...,0.000000,0.120743,0.000000,0.009335,0.000000,0.074976,0.088855,0.0,0.076203,0.011622
4,0.214442,0.066366,0.108787,1.000000,0.0,0.138000,0.117130,0.176688,0.231010,0.105819,...,0.219448,0.219436,0.082645,0.025090,0.095803,0.144622,0.243688,0.0,0.094962,0.049976
6,0.000000,0.000000,0.000000,0.000000,1.0,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5760,0.112011,0.000000,0.074976,0.144622,0.0,0.013749,0.080679,0.044892,0.133556,0.016102,...,0.176590,0.048857,0.117196,0.019998,0.009778,1.000000,0.066280,0.0,0.045822,0.017427
5766,0.238505,0.044879,0.088855,0.243688,0.0,0.151534,0.197204,0.218237,0.195913,0.109723,...,0.000000,0.311642,0.068307,0.044437,0.121670,0.066280,1.000000,0.0,0.028001,0.035405
5895,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.0,0.000000,0.000000


We will take a case for user with user_id 1

In [37]:
# Showing books that have been rated by `user_id=1`

user_id = 1

user_rating = ratings_matrix.T[user_id]
user_rating

,1
book_id,
1,NaN
2,NaN
3,NaN
4,5.0
5,NaN
...,...
9991,NaN
9995,NaN
9997,NaN


In [38]:
# Showing a list of users who have similar pattern to `user_id=1`

sim_users = df[user_id].sort_values(ascending=False)[1:]
sim_users.head(10)

,1
user_id,
5458,0.350813
4897,0.331543
3194,0.317092
4074,0.307814
571,0.297117
2181,0.296766
5167,0.291967
1423,0.291484
4546,0.290646


In [39]:
# Show the top 5 users similar to `user_id=1` along with their ratings for the first 5 books

top_5_user = ratings_matrix.loc[[5458, 4897, 3194, 4074, 571]]
top_5_user.iloc[:, :5]

book_id,1,2,3,4,5
user_id,,,,,
5458,NaN,NaN,NaN,NaN,NaN
4897,NaN,NaN,NaN,5.0,3.0
3194,NaN,NaN,NaN,5.0,4.0
4074,NaN,3.0,3.0,4.0,3.0
571,NaN,NaN,NaN,5.0,2.0


After you have determined a list of users similar to a user U, you need to calculate the rating R that U would give to a certain book.

<img src="https://files.realpython.com/media/weighted_rating.06ba3ea506b6.png" width=400></img>

In the above formula, every rating is multiplied by the similarity factor of the user who gave the rating. The final predicted rating by user U will be equal to the sum of the weighted ratings divided by the sum of the weights. With a weighted average, you give more consideration to the ratings of similar users in order of their similarity.

To make you easy to understand the codes below, the formula be written as:

```
predicted_rating = weighted_sum / total_sim
```

where `weighted_sum = sum(sim * ratings_matrix.loc[sim_user, book_id])`

In [40]:
# Finding ratings for books that `user_id=1` has not read

%%time

recommend = []

for book_id in ratings_matrix.columns: #Access all book_id on ratings_matrix dataframe
  if pd.isna(user_rating[book_id]): #Check if there's no rating to a book_id by user 1, it will be taken to the recommendation candidates list
    weighted_sum = 0 #initial weighted sum of rating
    total_sim = 0 #initial total similarity of user 1

    for sim_user, sim in sim_users.items(): #access all users and the similarity with user 1
      if pd.notna(ratings_matrix.loc[sim_user, book_id]): #Only rating by the users that have similarity with user 1 included
        weighted_sum += sim * ratings_matrix.loc[sim_user, book_id] #sum of similarity of each users with user 1 * rating that they gave to a book
        total_sim += sim #calculate all similarity

    if total_sim > 0: #If there are similarity among users and user 1, so we can calculate the predicted rating to a book by user 1
      recommend.append((book_id, weighted_sum/total_sim))

recommend.sort(key=lambda x: x[1], reverse=True)

CPU times: user 3min 12s, sys: 207 ms, total: 3min 12s
Wall time: 3min 23s


In [41]:
# Show the top 10 recommended books for `user_id=1`.

recommend[:10]

[(1781, np.float64(5.000000000000001)),
 (4099, np.float64(5.000000000000001)),
 (5788, np.float64(5.000000000000001)),
 (6435, np.float64(5.000000000000001)),
 (7060, np.float64(5.000000000000001)),
 (7839, np.float64(5.000000000000001)),
 (7846, np.float64(5.000000000000001)),
 (8286, np.float64(5.000000000000001)),
 (8947, np.float64(5.000000000000001)),
 (9727, np.float64(5.000000000000001))]

In [42]:
# Show the details of the 10 recommended books

list_of_recommended_books = [item[0]-1 for item in recommend[:10]]
books.loc[list_of_recommended_books][['book_id', 'original_title']]

,book_id,original_title
1780,1781,Don't Let the Pigeon Drive the Bus
4098,4099,The Heroin Diaries: A Year in the Life of a Sh...
5787,5788,ABNKKBSNPLAKo?! (Mga Kwentong Chalk ni Bob Ong)
6434,6435,The Making of the Atomic Bomb
7059,7060,Everyday Food: Great Food Fast
7838,7839,The Hidden Messages in Water
7845,7846,The Sight
8285,8286,On Food and Cooking: The Science and Lore of t...
8946,8947,Once a Runner
9726,9727,Heretic


Wrap up to a function

In [43]:
# Creating a recommender system function for all users

def rec_memory(user, top_number):
  user_rating = ratings_matrix.T[user]
  sim_users = df[user].sort_values(ascending=False)[1:]

  recommend = []

  for book_id in ratings_matrix.columns:
    if pd.isna(user_rating[book_id]):
      weighted_sum = 0
      total_sim = 0

      for sim_user, sim in sim_users.items():
        if pd.notna(ratings_matrix.loc[sim_user, book_id]):
          weighted_sum += sim * ratings_matrix.loc[sim_user, book_id]
          total_sim += sim

      if total_sim > 0:
        recommend.append((book_id, weighted_sum/total_sim))

  recommend.sort(key=lambda x: x[1], reverse=True)
  return recommend[:top_number]

If we take only top 5 to recommend user 10, so the list will be:

In [44]:
# Show the top 5 recommended books for `user_id=10`.
%%time

recommended_books = rec_memory(10, 5)
recommended_books

CPU times: user 3min 2s, sys: 167 ms, total: 3min 2s
Wall time: 3min 4s


[(4558, np.float64(5.000000000000002)),
 (153, np.float64(5.000000000000001)),
 (1803, np.float64(5.000000000000001)),
 (2836, np.float64(5.000000000000001)),
 (3225, np.float64(5.000000000000001))]

In [45]:
# Show the details of the 10 recommended books

list_of_recommended_books = [item[0]-1 for item in recommended_books]
books.loc[list_of_recommended_books]

,book_id,goodreads_book_id,best_book_id,work_id,books_count,isbn,isbn13,authors,original_publication_year,original_title,...,ratings_count,work_ratings_count,work_text_reviews_count,ratings_1,ratings_2,ratings_3,ratings_4,ratings_5,image_url,small_image_url
4557,4558,281512,281512,273545,48,849905702,9.780850e+12,Max Lucado,2000.0,NaN,...,19028,19443,271,221,431,2286,5473,11032,https://s.gr-assets.com/assets/nophoto/book/11...,https://s.gr-assets.com/assets/nophoto/book/50...
152,153,1582996,1582996,1575860,125,1416914293,9.781417e+12,Cassandra Clare,2008.0,City of Ashes,...,541117,586008,23643,6135,20075,92626,192934,274238,https://images.gr-assets.com/books/1432730356m...,https://images.gr-assets.com/books/1432730356s...
1802,1803,395871,395871,385368,33,006114097X,9.780061e+12,L.J. Smith,2007.0,The Awakening and the Struggle,...,68651,72227,3856,3663,8247,18819,19054,22444,https://images.gr-assets.com/books/1393837129m...,https://images.gr-assets.com/books/1393837129s...
2835,2836,71865,71865,1253352,78,439829100,9.780440e+12,John Marsden,1993.0,"Tomorrow, When the War Began",...,36501,40883,2574,718,1937,7428,15236,15564,https://s.gr-assets.com/assets/nophoto/book/11...,https://s.gr-assets.com/assets/nophoto/book/50...
3224,3225,634583,634583,620871,21,8173711461,9.788174e+12,"A.P.J. Abdul Kalam, Arun Tiwari",1999.0,Wings of Fire: An Autobiography,...,24450,26675,848,258,785,4068,10002,11562,https://images.gr-assets.com/books/1295670969m...,https://images.gr-assets.com/books/1295670969s...
